### PR TITLE
SO(3)-equivariant shear head: tangent-frame vector prediction

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**SO(3)-aware shear head: replace the scalar MLP output head for wsy/wsz with a cross-product-based equivariant vector output will improve wall-shear axis prediction.**

Wall shear stress is a **vector quantity** — it transforms under rotation just as a tangential force vector does. The current model predicts wsy and wsz as independent scalar channels, discarding the physical constraint that the shear vector lies in the tangent plane at each surface point and transforms equivariantly under rotations. This is a systematic inductive bias violation.

**The fix:** instead of predicting [wsx, wsy, wsz] as 3 independent scalars, predict the shear vector as a linear combination of the local tangent frame basis vectors {t1, t2, n} where n is the surface normal. The output head predicts two scalar coefficients [α, β] in the tangent plane, and the shear vector is reconstructed as `w = α*t1 + β*t2`. This enforces that:
1. The shear vector lies in the tangent plane (normal component is zero, physically correct for wall shear)
2. The prediction is equivariant under rotations of the mesh

This approach is inspired by "Frame Averaging for Equivariant Shape Space Learning" (Puny et al. 2022) and tangent-space vector field prediction in geometric DL.

**Note on tangent frames:** The DrivAerML dataset includes surface normal vectors for each point. Tangent vectors {t1, t2} can be constructed from the normals using Gram-Schmidt. Check the dataset/dataloader to see if normals are already available as features.

### 1. Check available point features

First, inspect the dataset to confirm what per-point features are available:

```python
# In your debug script or train.py, print a sample batch
print(batch.keys())
print(batch['x'].shape)  # node features
# Check if normals are in the node features or available as a separate field
```

If surface normals are available (look for `normal_x`, `normal_y`, `normal_z` in the dataset), proceed with the equivariant head. If not, normals can be estimated from the mesh topology (cross-product of edge vectors at each vertex).

### 2. Build local tangent frames

```python
def build_tangent_frame(normals):
    """Build orthonormal tangent frame {t1, t2} from surface normals.
    normals: (N, 3) surface normal vectors (unit)
    Returns: t1 (N, 3), t2 (N, 3)
    """
    # Gram-Schmidt: pick an arbitrary non-parallel reference vector
    ref = torch.zeros_like(normals)
    ref[:, 0] = 1.0  # x-axis
    # Where normals are nearly parallel to x-axis, use y-axis instead
    parallel_mask = (normals[:, 0].abs() > 0.9)
    ref[parallel_mask, 0] = 0.0
    ref[parallel_mask, 1] = 1.0
    
    # t1 = normalize(ref - (ref · n) * n)
    t1 = ref - (ref * normals).sum(-1, keepdim=True) * normals
    t1 = F.normalize(t1, dim=-1)
    # t2 = n × t1
    t2 = torch.cross(normals, t1, dim=-1)
    return t1, t2
```

### 3. Modify the output head for wall shear

Replace the current wsy/wsz scalar prediction with tangent-frame coefficient prediction:

```python
class EquivariantShearHead(nn.Module):
    def __init__(self, hidden_dim):
        super().__init__()
        # Predict 2 tangent-frame coefficients instead of 3 scalars
        self.proj = nn.Linear(hidden_dim, 2)  # [alpha, beta] for t1, t2
    
    def forward(self, h, t1, t2):
        # h: (N, hidden_dim) node hidden states
        # t1, t2: (N, 3) tangent frame vectors
        coeffs = self.proj(h)  # (N, 2)
        # Reconstruct shear vector in 3D
        w = coeffs[:, 0:1] * t1 + coeffs[:, 1:2] * t2  # (N, 3)
        return w  # predicted [wsx, wsy, wsz]
```

Add a CLI flag `--equivariant-shear-head` (boolean, default False). When enabled, replace the shear output head with `EquivariantShearHead`.

### 4. Training run

```bash
cd target/ && torchrun --standalone --nproc-per-node=4 train.py \
  --model-layers 5 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --fourier-pe \
  --no-use-ema \
  --lr 3e-4 \
  --lr-cosine-t-max 50 \
  --no-compile-model \
  --equivariant-shear-head \
  --kill-thresholds "500:train/loss<5,178000:val_primary/abupt_axis_mean_rel_l2_pct<10" \
  --wandb-group bengio-wave20-equivariant-shear \
  --wandb-name chihiro/so3-equivariant-shear \
  --agent chihiro
```

**If normals are not available in the dataset:** implement a fallback that estimates normals from mesh connectivity (cross-product of two edge vectors per vertex). Report whether normals were dataset-provided or estimated.

**If ep10 val_abupt > 9.0%:** the tangent frame construction may be incorrect — check that t1, t2 are orthogonal to n and to each other, and that the reconstruction loss is computing against [wsx, wsy, wsz] correctly.

## Baseline

Current best (PR #174, alphonse, val metrics at ep~45.3, step 807,025, run `vu4jsiic`):

| Metric | Current Best (val) | AB-UPT Target | Gap |
|--------|-------------------|--------------|-----|
| `val_primary/abupt_axis_mean_rel_l2_pct` | **6.9549** | 4.51 | 2.44pp |
| `val_primary/surface_pressure_rel_l2_pct` | 4.5644 | 3.82 | 0.74pp |
| `val_primary/volume_pressure_rel_l2_pct` | **3.9361** ✓ | 6.08 | solved |
| `val_primary/wall_shear_y_rel_l2_pct` | 8.7345 | 3.65 | **5.08pp** |
| `val_primary/wall_shear_z_rel_l2_pct` | 10.5766 | 3.63 | **6.95pp** |

**Target:** wsy below 8.0% and wsz below 10.0% at ep50, indicating the equivariant head better captures shear physics.

Reproduce baseline:
```bash
cd target/ && torchrun --standalone --nproc-per-node=4 train.py \
  --model-layers 5 --model-hidden-dim 256 --model-heads 4 \
  --fourier-pe --no-use-ema --lr 3e-4 --lr-cosine-t-max 50 \
  --no-compile-model --wandb-group bengio-wave9
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Acknowledged — hypothesis received. SO(3)-equivariant shear head with tangent-frame vector prediction is a principled physics-motivated approach — wall shear is a tangential surface vector and standard L2 regression on Cartesian components doesn't respect this geometry. Current baseline to beat: **val_abupt = 6.9549%** (PR #174, run `vu4jsiic`, ep~45.3).

**Kill gates (abupt):** EP5 > 13%, EP10 > 10.5%, EP15 > 8.5%, EP20 > 8.0%.

Key thing to watch: does wsy/wsz improve relative to surface_pressure? If the equivariant head helps wall-shear without hurting pressure, that's the signal we want. Report full per-axis breakdown.

Post your W&B run ID(s) and first val checkpoint when training starts.

**Comment by morganmcg1** [comment]
STUDENT chihiro: This is a follow-up on your assigned PR. Please confirm you have started your training run and post your W&B run ID(s). The baseline to beat is **val_abupt = 6.9549%** (PR #174, run `vu4jsiic`). Post ep1 metrics as soon as available and apply standard kill gates (ep5: >13%, ep10: >10.5%, ep15: >8.5%, ep20: >8.0%).

**Comment by morganmcg1** [comment]
ADVISOR (infra-block escalation, 2026-05-03): Pod `senpai-bengio-chihiro` is confirmed blocked by a zombie `train.py` process from a prior wave. This is a known infrastructure issue (#466) — **you are not at fault**. The human infrastructure team has been notified and must SIGKILL the zombie process and restart the deployment.

Once the pod restarts and is free, launch your SO3-equivariant shear head run immediately per the PR instructions. Post your W&B run ID(s) and ep1 metrics as soon as available. No further action needed from you until the pod is freed.
